velox/functions/prestosql/WordStem.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <libstemmer.h>
+
+#include "velox/functions/Udf.h"
+#include "velox/functions/lib/string/StringImpl.h"
+
+namespace facebook::velox::functions {
+
+namespace {
+/// Wrap the sbstemmer library and use its sb_stemmer_stem
+/// to get word stem
+class Stemmer {
+ private:
+  sb_stemmer* sbStemmer_;
+  Stemmer(sb_stemmer* stemmer) : sbStemmer_(stemmer) {}
+
+ public:
+  ~Stemmer() {
+    sb_stemmer_delete(sbStemmer_);
+  }
+
+  /// Get a Stemmer from the the map stored in thread local storage
+  /// or create a new one if it doesn't exist. Return nullptr if the
+  /// specified lang is not supported.
+  static Stemmer* getStemmer(const char* lang) {
+    thread_local std::map<std::string, std::unique_ptr<Stemmer>> stemmers;
+    if (auto found = stemmers.find(lang); found != stemmers.end()) {
+      return found->second.get();
+    }
+    Stemmer* stemmer = nullptr;
+    // Only support ASCII and UTF-8
+    if (auto sbStemmer = sb_stemmer_new(lang, "UTF_8")) {
+      stemmer = new Stemmer(sbStemmer);
+      stemmers[lang] = std::unique_ptr<Stemmer>(stemmer);
+    }
+    return stemmer;
+  }
+
+  /// Get the word stem or NULL if out of memory
+  const char* stem(const std::string& input) {
+    return (const char*)(sb_stemmer_stem(
+        sbStemmer_,
+        reinterpret_cast<unsigned char const*>(input.c_str()),
+        input.length()));
+  }
+};
+} // namespace
+
+/// word_stem function
+/// word_stem(word) -> varchar
+///     return the stem of the word in the English language
+/// word_stem(word, lang) -> varchar
+///     return the stem of the word in the specificed language
+///
+/// It uses the snowball stemmer library to calculate the stem.
+/// https://snowballstem.org
+/// It provides Java implementation which is used in Presto as well
+/// as C implementation. Therefore, both Presto and Prestimissio
+/// would have the same word stem results.
+template <typename TExec>
+struct WordStemFunction {
```

Comment:
@yhwang All you need to do is the following:

1- Add member variable to WordStemFunction to store a map of stemmers.
2- Move getStemmer method into WordStemFunction struct as a private method.

Reply:
When I move getStemmer into WordStemFunction, it moves the `std::map<std::string, std::unique_ptr<Stemmer>> stemmers` together. Then wait for the thread-local decision, right?

Reply:
> Then wait for the thread-local decision, right?

Not sure if we need to wait. Let's remove 'thread-local' and make the map a regular private member of the WordStemFunction struct.

FYI, in case you haven't seen these already: https://facebookincubator.github.io/velox/develop/scalar-functions.html